### DISCUSSION TITLE
Add color range legend, record video, config

### DISCUSSION THREAD

**Discussion by AKYD** [tomshaffner/PiThermalCam#3] [Ideas]
Hello,

Very nice project - ran it on RPi2 and works ok :)

I do have a couple of suggestions:

- add a color range legend (right side, toggle on/off?) like in the image below:

![image](https://user-images.githubusercontent.com/2852960/110941722-77be3880-8341-11eb-9b3f-273b2446c883.png)

- add a Record video option similar to the "Save" option (maybe with pre-configured duration)

- add config so startup settings are the same on each run (C vs F units for temp for example)

- Capture normal image from camera and overlay the thermal values on top of the normal image (Flir calls this [MSX](https://www.flir.com/discover/professional-tools/what-is-msx/))

I'll try to have a look at the code and see if I figure out how to implement some of these.

Cheers

**Comment by Gatada** [comment]
Is that how the camera works? Would be great if it is. I thought the colours didn't correlate to an actual temperature, but more to the colour difference? So, the hottest source is always the same colour.

Disclaimer: not tried it yet, but I hope to when time allows.

**Reply by AKYD** [reply]
As far as I understand it, the hottest color corresponds with the higher temperature recorded at that moment and similarly for the coldest color so it would be a matter of creating `(hottestColor-coldestColor) / 10` color areas I guess for 10 color ranges

Did a quick test by lighting a match and the hottest color changed to something hotter :)

**Reply by tomshaffner** [reply]
These sound like great ideas but for the moment I'm caught up with other things and won't have time to commit here for a while; you're welcome to add them yourself and submit a pull request! A few thoughts:

Config: My hope was that many of the settings are now simple enough to change with the buttons or keystrokes that things like which unit you're using at startup become irrelevant. If you only need to hit "T" to change it, it's less of an issue. A config for this wouldn't be hard to make though, I actually had a config in the original version but stripped it when the number of parameters got too small. For your own runs it might be easier simply to change the defaults in the code, but this change is certainly doable.

Normal Image Overlay: To do this we'd need to implement a full and equivalent normal camera in addition to the thermal one and then, if the overlay were to be exact, I'd have to map the two images to one another; that's actually an incredibly complex process. For my purposes I thought it made more sense just to take two pictures and line them up myself. I'd imagine FLIR can do that easily because phones have a normal camera built in and because they can know, based on the phone type, the location of that camera in comparison to the thermal one and that the angle of both is the same. The Pi implementation though would be much more variable so I'd think that would be incredibly hard to do in a generalizable way. That and the quality here is lower making that lining up much harder. If you start to dig into the code yourself I'd suggest saving this for a time when you've got a lot of time; it would be pretty cool but I don't think it would go quick.

The record video seems like the point here that would be the biggest value add; that would be a great addition. It's not something I've done before, I'd need to look into the process for that. If you don't do a PR for that though this might be where I focus my efforts if/when I get back to this project.

The legend point is a good one. In practice though, the temperatures the camera gives are not absolute but relative. That means the difference between 20 and 50 should in fact be 30 degrees, but the 20 and 50 might not be correct. Apparently getting the actual real temperatures requires a calibration process, and I'm not sure it holds between uses, so it's complex. For my purposes the value was more about the relative temperatures anyway; seeing where a window or a corner of a wall is much colder than the surrounding wall, etc. And for that purpose I could also see _how much_ colder even without knowing the absolute temperature.

Of course, the legend could still be good for having a deeper sense of the gradations between areas you're dealing with, but I also just didn't know how to implement it easily in the CV2 system. I believe it's still in the matplotlib approach if you want to use it instead and try and tweak more to get your FPS up; check the original article on that for more detail. Wherever such exists though we'd need to emphasize more clearly that the temperatures are relative, not absolute.
